lib/SILOptimizer/Mandatory/DiagnoseVarUsage.cpp

```diff
@@ -0,0 +1,1617 @@
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a diagnostic pass to diagnose unused, never read, and
+// never mutated values.
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "diagnose-var-usage"
+#include "swift/AST/DiagnosticEngine.h"
+#include "swift/AST/DiagnosticsSIL.h"
+#include "swift/AST/ASTWalker.h"
+#include "swift/AST/Expr.h"
+#include "swift/AST/Pattern.h"
+#include "swift/AST/Stmt.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SIL/SILValue.h"
+#include "swift/SIL/SILArgument.h"
+#include "swift/SIL/ApplySite.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+
+using namespace swift;
+
+//===----------------------------------------------------------------------===//
+//                          MARK: Utility Functions
+//===----------------------------------------------------------------------===//
+
+/// Returns the \c VarDecl this instruction belongs to or \c nullptr
+static VarDecl *getVarDecl(SILInstruction *I) {
+  if (auto Loc = I->getLoc()) {
+    if (auto *VD = Loc.getAsASTNode<VarDecl>()) {
+      return VD;
+    }
+  }
+  return nullptr;
+}
+
+/// Returns whether this instruction belongs to a setter parameter \c newValue
+static bool isSetterParam(SILInstruction *I) {
+  auto *VD = getVarDecl(I);
+  if (!VD) {
+    return false;
+  }
+  if (auto PD = dyn_cast<ParamDecl>(VD)) {
+    if (auto AD = dyn_cast<AccessorDecl>(VD->getDeclContext())) {
+      if (AD->getAccessorKind() == AccessorKind::Set) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+/// Returns if the SILInstruction if it is part of a VarDecl we want to
+/// track for diagnostics.
+static bool shouldTrackVarDecl(SILInstruction *I) {
+  if (auto *VD = getVarDecl(I)) {
+    
+    if (!VD->shouldDiagnoseUsage()) {
+      return false;
+    }
+    
+    if (VD->isCaptureList()) {
+      return true;
+    }
+    
+    if (isSetterParam(I)) {
+      return true;
+    }
+    
+    // If the variable is implicit, ignore it.
+    if (VD->isImplicit() || VD->getLoc().isInvalid())
+      return false;
+    
+    // If the variable is computed, ignore it.
+    if (!VD->hasStorage())
+      return false;
+    
+    // If the variable was invalid, ignore it and notice that the code is
+    // malformed.
+    if (VD->isInvalid()) {
+      return false;
+    }
+    
+    // If the variable is already unnamed, ignore it.
+    if (!VD->hasName() || VD->getName().str() == "_")
+      return false;
+    
+    return (VD->getDeclContext()->isLocalContext() && !isa<ParamDecl>(VD));
+  }
+  return false;
+}
```

Comment:
Maybe coalesce these to some extent?

```suggestion
static bool shouldTrackVarDecl(SILInstruction *I) {
  auto *VD = getVarDecl(I);
  if (!VD || !VD->shouldDiagnoseUsage()
      || !VD->hasName() || VD->getName().str() == "_"
      || VD->isImplicit() || VD->getLoc().isInvalid()
      || !VD->hasStorage() || VD->isInvalid()) {
      return false;
  }
  return VD->isCaptureList() || isSetterParam(I) ||
    (VD->getDeclContext()->isLocalContext() && !isa<ParamDecl>(VD));
}
```

I'm not sure if the comments are adding much, they're not describing the why but only what the code is doing.

Also, this part:
```
    // If the variable was invalid, ignore it and notice that the code is
    // malformed.
```
seems like a copy-paste error, as there is no `sawError = true` here.